go/arrow/array/float16.go

```diff
@@ -0,0 +1,84 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package array
+
+import (
+	"fmt"
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/numeric"
+	"strings"
+)
+
+// A type which represents an immutable sequence of Float16 values.
+type Float16 struct {
+	array
+	values []numeric.Float16
+}
+
+func NewFloat16Data(data *Data) *Float16 {
+	a := &Float16{}
+	a.refCount = 1
+	a.setData(data)
+	return a
+}
+
+func (a *Float16) Value(i int) numeric.Float16 { return a.values[i] }
+
+func (a *Float16) Float32Value(i int) float32 { return a.values[i].Float32() }
+
+func (a *Float16) Values() []numeric.Float16 { return a.values }
+
+func (a *Float16) Float32Values() []float32 {
```

Comment:
this one carries perhaps a bit more weight, but could also be tackled by the `float16` package:
if we were to add a type in that package, say `Slice` or `Floats`, one could define a method on that type to convert to a slice of `float32`:
```go
type Floats []Float16
func (fs Floats) Float32s() []float32 { ... }
```

and store a value of this type inside the `Float16` array.

WDYT?

Reply:
I'm not sure... We don't have a specific type for other logical numeric slice types (date and time)

Reply:
OK, let's remove Float32 version for now, unless we see any need in future